design/authn_azure/authn_azure_solution_design.md

```diff
@@ -0,0 +1,357 @@
+| 2  | Policy is loaded w/o "provider-uri" variable or value is missing                          | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (RoleNotAuthorizedOnWebservice) is returned and logged | Integration               |            |
+| 3  | Policy is loaded w/o defining authenticator webservice                                    | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (WebserviceNotFound) is returned and logged            | Integration               |            |
+| 4  | An incorrectly/undefined provider-uri (Conjur couldn't connect to provider)               | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (ProviderDiscoveryTimeout) is returned and logged      | Integration               |            |
+| 5  | Conjur can make contact with provider but couldn't confirm token signature                | VM sends authentication request to Conjur with invalid Azure token  | 500 Internal Server Error response with an error body with the relevant error message (ProviderTokenInvalid) is returned and logged          | Integration               |            |
+| 6  | One or more annotations are missing from the host policy **                               | VM sends authentication request to Conjur with valid Azure token    | 500 Internal Server Error response with an error body with the relevant error message (MissingRequestParam) is returned and logged           | Integration               |            |
+| 7  | One or more fields in Azure token is missing                                              | VM sends authentication request to Conjur with invalid Azure token  | 500 Internal Server Error response with an error body with the relevant error message (TokenFieldNotFoundOrEmpty) is returned and logged     | Integration               |            |
+| 8  | One or more fields in the Azure token don't match the Conjur host annotations             | VM sends authentication request to Conjur with valid   Azure token  | 500 Internal Server Error response with an error body with the relevant error message (InvalidApplicationIdentity) is returned and logged    | Integration               |            |
+
+### Unit tests
+
+1. Validate that `validate_azure_token` properly decodes the of JWT token
+
+1. Validate that `validate_application_identity` accurately compares the Conjur host identity and Azure token correctly
+
+1. If `validate_azure_token` or `validate_application_identity` raise an error then it propagates to calling `Authentication::Authenticate` class
+
+### Performance
```

Comment:
1. How many sessions we can run in parallel ? need to test it
2. How we can declare pass for performance ? verify 3000 sessions and verify the average is 1 sec per call ?
3. Stability - how many requests we send ? do we verify during this test some failover and see recovery ?

Reply:
@eladkub 
1. We can start with ±20 the raise the bar in GA.
3. Yes, such scenarios should be tested but not necessarily with each and every authenticator. A scenario for this authenticator can be unavailability of the Azure AD or restart of Application VM.

Reply:
@InbalZilberman for stability tests do we want to measure requests over a few days?

Reply:
Also, 20 requests per vm?